### PR TITLE
Validate cpp_info.name

### PR DESCRIPTION
**Pull request by uilianries**
- Update Conan version for testing
- Add new hook which checks cpp_info.name is declared
- Add tests for hook KB-H040

Related issue: https://github.com/conan-io/conan/issues/6269#issuecomment-570195413

Closes https://github.com/conan-io/conan/issues/6001

/cc @Morwenn

Wiki:

#### **<a name="KB-H040">#KB-H032</a>: "NO TARGET NAME"**

According the Conan issue [#6269](https://github.com/conan-io/conan/issues/6269), the attribute `cpp_info.name` should be avoided for Conan Center Index in favor of `cpp_info.names["cmake_find_package"]` and `cpp_info.names["cmake_find_multi_package"]`.

### CODE REVIEW

**Comment by danimtb** [comment]
Please add the description of the error KB in the description of this PR

**Comment by danimtb** [comment]
@jgsogo there is a nightly job doing just `conan export` over all the recipes with hooks activated in order to test them. I think that should be enough, right?

**Comment by jgsogo** [comment]
Yes, that one is enough. The idea is to check if using `cpp_info.name` is the way to go for some recipe and then we cannot activate this hook worlwide.

**Comment by danimtb** [comment]
I see some packages that are still using cpp_info.name: bzip2, entt and openssl. I would like to have those updated first before merging this

**Comment by uilianries** [comment]
> I see some packages that are still using cpp_info.name: bzip2, entt and openssl. I would like to have those updated first before merging this

OpenSSL: https://github.com/conan-io/conan-center-index/pull/604
EnTT: https://github.com/conan-io/conan-center-index/pull/673
Bzip2: https://github.com/conan-io/conan-center-index/pull/672

**Comment by danimtb** [comment]
@uilianries https://github.com/conan-io/hooks/pull/142#issuecomment-571943400

**Comment by danimtb** [comment]
Please update this PR when possible and add the description of the KB in the description of this PR

**Comment by uilianries** [comment]
@danimtb Conflicts has been fixed. Also, I've added the wiki description, please take a look. Thanks for pinging me.

File: hooks/conan-center.py [hunk 1]
```diff
@@ -359,6 +360,16 @@ def test(out):
...
+    @run_test("KB-H040", output)
+    def test(out):
+        if "self.cpp_info.name =" in conanfile_content:
+            out.error("Conanfile should not contain 'self.cpp_info.name'."
+                      " Use 'cpp_info.names' instead.")
+
+        if "self.cpp_info.names['cmake']" in conanfile_content or \
+           'self.cpp_info.names["cmake"]' in conanfile_content:
+            out.error("Conanfile should not contain 'self.cpp_info.names['cmake']'."
+                      " Use 'cmake_find_package' and 'cmake_find_package_multi' instead.")
```
Comment (jgsogo) [inline hooks/conan-center.py]: I think that the reason to avoid `self.cpp_info.names['cmake']` is because it should match the name of the package

**Comment by jgsogo** [comment]
> According the Conan issue [#6269](https://github.com/conan-io/conan/issues/6269), the attribute `cpp_info.name` has been deprecated in favor of `cpp_info.names["cmake_find_package"]` and `cpp_info.names["cmake_find_multi_package"]`.

I wouldn't say that `cpp_info.name` is deprecated, just say that we prefer not to use it in Conan Center 😉

File: hooks/conan-center.py (hunk 1 above)
Comment (uilianries) [reply hooks/conan-center.py]: okay, but what message should we use instead?

**Comment by uilianries** [comment]
> > According the Conan issue [#6269](https://github.com/conan-io/conan/issues/6269), the attribute `cpp_info.name` has been deprecated in favor of `cpp_info.names["cmake_find_package"]` and `cpp_info.names["cmake_find_multi_package"]`.
> 
> I wouldn't say that `cpp_info.name` is deprecated, just say that we prefer not to use it in Conan Center wink

I just updated it now, thanks for reviewing

File: hooks/conan-center.py (hunk 1 above)
Comment (jgsogo) [reply hooks/conan-center.py]: Something like: _"Conan Center uses the name of the package for `cmake` generator, do not change it here"_, nevertheless, if this is needed, would it be possible to add the package to a whitelist or something similar?

**Comment by Morwenn** [comment]
As mentioned in conan-io/conan-center-index#787 the restriction that applies to `cmake` should probably apply to `cmake_multi` too for consistency.

File: hooks/conan-center.py (hunk 1 above)
Comment (uilianries) [reply hooks/conan-center.py]: Depending the package, is possible add it to a allow list. However, that hook will prevent unsuspecting people.

**Comment by uilianries** [comment]
@jgsogo @Morwenn I just update the branch, please, review again.

**Comment by jgsogo** [comment]
@Morwenn, I think that everything that applies to 'cmake' should apply to 'cmake_multi' too (at least in 'cpp_info'). And the same for 'cmake_find_package' and 'visual_studio', and any other '_multi' out there. Probably someone should open the issue :D

File: hooks/conan-center.py [hunk 2]
```diff
@@ -359,6 +360,19 @@ def test(out):
                 author = '"%s"' % author
             out.error("Conanfile should not contain author. Remove 'author = {}'".format(author))
 
+    @run_test("KB-H040", output)
+    def test(out):
+        if "self.cpp_info.name =" in conanfile_content:
+            out.error("CCI uses the name of the package for cmake generator."
```
Comment (Morwenn) [inline hooks/conan-center.py]: Won't the first line confuse users who didn't try to use `cpp_info.name` for the CMake generators?

File: hooks/conan-center.py [hunk 3]
```diff
@@ -359,6 +360,19 @@ def test(out):
...
+                      " Conanfile should not contain 'self.cpp_info.name'."
+                      " Use 'cpp_info.names' instead.")
+
+        for generator in ["cmake", "cmake_multi"]:
+            if "self.cpp_info.names['{}']".format(generator) in conanfile_content or \
+               'self.cpp_info.names["{}"]'.format(generator) in conanfile_content:
+                out.error("CCI uses the name of the package for cmake generator. "
```
Comment (Morwenn) [inline hooks/conan-center.py]: Maybe replace `cmake` in this string with the value of `generator`?

File: hooks/conan-center.py (hunk 2 above)
Comment (uilianries) [reply hooks/conan-center.py]: done, updated

File: hooks/conan-center.py (hunk 3 above)
Comment (uilianries) [reply hooks/conan-center.py]: done, updated

**Comment by Morwenn** [comment]
@uilianries It's good for me, thanks for everything 😄 

With that in it's probably the better answer we can give to my original issue conan-io/conan#6001 right now. It will probably be good enough to close the issue since I don't believe there is a better solution to the problem right now.

**Comment by uilianries** [comment]
@Morwenn thanks a lot for reviewing!

**Pull request merged**